core/standards.md

```diff
@@ -0,0 +1,69 @@
+# Core Smart Contracts
+
+## Definition of Core
+
+A core contract is one that:
+
+1. Any contract that is part of the [dss repo](https://github.com/makerdao/dss)
+2. will have ward status on or other privileged access to any other core contract (i.e. `Vat.rely(<new contract address>)`)
+3. will rely on the Pause Proxy as a permissioned actor. If the pause proxy has to directly interact with this contract then there is additional security and process work that needs to be done.
+4. will be included in the chainlog as a canonical “Maker Contract”. Even if the contract does not have elevated permissions in the system, a failure of a contract in the chainlog could result in cost (bug bounties) or reputation damage for the protocol.
+5. If a contract is put forward to a business partner of the DAO, by DAO members, close consideration should be given to whether it would be perceived as core Maker brand by those partners.
+
+## Protocol Engineering Review Process for Core Contracts
+
+All Core contracts need to go through a risk adjusted review process.  Risk should be evaluated at the beginning of the project and continue to be re-evaluated through the engineering process.
+
+Our review process at a minimum requires two team member reviews and approvals.  However for higher risk contracts may also involve team Red Team reviews and audits by external auditors.
+
+Core contracts should follow established coding patterns including repo and contract structures, variable naming patterns (yes Daiwanese is a standard for our team) and auth patterns.
+
+Whenever possible, utilize an existing solution instead of building a new contract. MakerDAO is a very complex project and each new contract or derivation increases the cognitive and maintenance overhead of the entire protocol. By limiting the number of moving parts we reduce the risk of unexpected behavior.
+
+This repo attempts to document our standards, other sources include (non-exhaustive list):
+
+- https://github.com/makerdao/mips-best-practices
+
+## Coding standards we expect to see in core contracts
+
+### Permissions
+
+Permissioning in the contract should be handled by `wards` and the `wards` mapping should occupy memory slot `0x0`
+
+### Contract ordering
+
+Following and expanding on [Solidity Order of layout recommendations](https://docs.soliditylang.org/en/stable/style-guide.html#order-of-layout) our contracts are generally laid out in this way:
+
+1. State variables (including wards and math)
+
+    a. **note:** wards should be memory slot `0x0`
+2. Events
+3. Modifiers
+4. Internal Math variables and functions
+5. Other Internal functions
+6. Wards functions
+7. External functions
```

Comment:
```suggestion
7. Fileable functions
8. External functions
```

Reply:
Fileable functions are part of wards functions.

Reply:
this should be specified then I guess

Reply:
I guess "wards functions" is a synonym of "authed functions". Both are fine by me so whatever you guys prefer.

Reply:
If the goal of the "Standards" section is to have clear guidelines for core contract development, this should be expressed clearly in it's meaning to avoid confusion or ambiguity.

Reply:
Fileables is part of the contract layout structure, with it's own pattern and I think that's worth documenting, besides the property of being authed.

Reply:
I think is clearly described as it is. If you think the opposite is fine to change to `authed functions`.
It is also fine if you want to split the point in three (`rely/deny`, `file` and other `authed functions`). I don't think is necessary that much verbosity but not opposed either if it is the preferred way for the rest.

Reply:
So, this is a good point, I was referring specifically to the ward related functions **not** any function with `auth`.  I don't think we have typically grouped `auth` functions at the top have we?  I'll take gonza's suggestion on outlining the three types since it looks like the current wording left room for interpretation.

Reply:
how does this sound to you all: https://github.com/makerdao/pe-checklists/pull/2/commits/9f4de9dbf421aba827e5ea575ce6b13de0d95397

Reply:
I would use the `Fileable` terminology here, as it's a unique pattern.
Should we also include `authed functions` after the Fileable functions as suggested by @gbalabasquer ?

Reply:
I don't believe we have been putting `authed functions` at the top of `external` functions.  is that a standard somewhere? D3M has the external functions ordered differently

Reply:
I would follow `dss-vest` contract layout for reference on that, after fileables there are internal math and then contract-specific authed functions

Reply:
hmmm, i'm not sure dss-vest is the standard right now. it looks like its layout is:

  1. Variables
  2. Events
  3. _Public Getter Functions_
  4. Constructor 
  5. Modifiers
  6. Wards functions
  7. Fileable
  8. Internal Math functions
  9. One Auth function
  10. External Function (`vest(uint256)` and `vest(uint256, uint256)`
  11. Internal function (`_vest`)
  12. External Function (`accrued`)
  13. Internal function (`accrued` _<-- missing underscore_)
  14. etc

So if we want to declare that as the new standard i can put that here, but our more recent contracts (d3m, dss-teleport) don't follow parts of that

Reply:
I'm not saying `dss-vest` should be used as a standard but just extrapolating common patterns from dss-contracts

Reply:
the missing underscore I think relates to an overloaded function

Reply:
ok, sorry i misunderstood:

> I would follow dss-vest contract layout for reference on that, 

so are you happy with this as the rough standard (also keep in mind this doc specifies "general guidelines" that we can deviate from when justified):

1. State variables (including wards and math)
    a. constants then immutables then storage variables
    b. wards should be storage slot `0x0`
2. Events
3. Constructor
4. Modifiers
5. Internal Math variables and functions
6. Wards functions (i.e. `rely`/`deny`)
7. Fileable functions
8. Other Internal functions
9. Other External functions

Reply:
looks good!
I would inlcude, following `dss-vest` Getters after Events

Reply:
> looks good! I would inlcude, following `dss-vest`, Getters after Events

for d3m we found that those getters were getting in the way and put them at the bottom, are you okay with that?

Reply:
I'm not strongly opposed but for a formalisation of that I would check with the team and maybe considering putting them after Events in a compact format.

Reply:
we also might want to add NatSpec as part of the best practises